finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,719 @@
+use crate::{
+    new_network::{
+        connection_manager::{Authentication, Multiaddr, SessionHandler},
+        DataCommand, PeerId, Protocol,
+    },
+    NodeCount, NodeIndex, SessionId,
+};
+use codec::{Decode, Encode};
+use std::{
+    collections::{HashMap, HashSet},
+    time::{Duration, Instant},
+};
+
+/// How many nodes we should query about unknown authorities in one go.
+const NODES_TO_QUERY: usize = 2;
+
+/// Messages used for discovery and authentication.
+#[derive(Clone, Debug, Encode, Decode)]
+pub enum DiscoveryMessage {
+    AuthenticationBroadcast(Authentication),
+    // Requests always contain own authentication, to avoid asymmetric trust.
+    Request(Vec<NodeIndex>, Authentication),
+    // Always assumed to contain only authentications for one session.
+    // Only authentications from the same session as the first present are guaranteed to be
+    // processed.
+    Authentications(Vec<Authentication>),
+}
+
+impl DiscoveryMessage {
+    pub fn session_id(&self) -> SessionId {
+        use DiscoveryMessage::*;
+        match self {
+            AuthenticationBroadcast((auth_data, _)) => auth_data.session(),
+            Request(_, (auth_data, _)) => auth_data.session(),
+            Authentications(auths) => match auths.get(0) {
+                Some((auth_data, _)) => auth_data.session(),
+                None => SessionId(0), // Broken message anyway, value doesn't matter.
+            },
+        }
+    }
+}
+
+/// Handles creating and responding to discovery messages.
+pub struct Discovery {
+    cooldown: Duration,
+    last_broadcast: HashMap<NodeIndex, Instant>,
+    last_response: HashMap<NodeIndex, Instant>,
+    requested_authentications: HashMap<NodeIndex, HashSet<NodeIndex>>,
+    // Used to rotate the nodes we query about unknown nodes.
+    next_query: NodeIndex,
+}
+
+type DiscoveryCommand = (DiscoveryMessage, DataCommand);
+
+fn authentication_broadcast(authentication: Authentication) -> DiscoveryCommand {
+    (
+        DiscoveryMessage::AuthenticationBroadcast(authentication),
+        DataCommand::Broadcast,
+    )
+}
+
+fn request(
+    missing_authorities: Vec<NodeIndex>,
+    authentication: Authentication,
+    peer_id: PeerId,
+) -> DiscoveryCommand {
+    (
+        DiscoveryMessage::Request(missing_authorities, authentication),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+fn response(authentications: Vec<Authentication>, peer_id: PeerId) -> DiscoveryCommand {
+    (
+        DiscoveryMessage::Authentications(authentications),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+impl Discovery {
+    /// Create a new discovery handler with the given response/broadcast cooldown.
+    pub fn new(cooldown: Duration) -> Self {
+        Discovery {
+            cooldown,
+            last_broadcast: HashMap::new(),
+            last_response: HashMap::new(),
+            requested_authentications: HashMap::new(),
+            next_query: NodeIndex(0),
```

Comment:
Wouldn't it be better to somehow randomize this point?

Reply:
Possibly, but I'm not convinced the added value is significant, while randomness is sorta annoying to deal with.

Reply:
Why? It doesn't have to be perfect randomness, I just think it can help with distribution of network traffic

Reply:
Added some randomness, not quite happy with the outcome, but it should help with the failure mode you mentioned to me.